crates/cli/tests/reference/static.js

```diff
@@ -0,0 +1,77 @@
+let wasm;
+export function __wbg_set_wasm(val) {
+    wasm = val;
+}
+
+
+function isLikeNone(x) {
+    return x === undefined || x === null;
+}
+
+function addToExternrefTable0(obj) {
+    const idx = wasm.__externref_table_alloc();
+    wasm.__wbindgen_export_1.set(idx, obj);
+    return idx;
+}
+
+const lTextDecoder = typeof TextDecoder === 'undefined' ? (0, module.require)('util').TextDecoder : TextDecoder;
+
+let cachedTextDecoder = new lTextDecoder('utf-8', { ignoreBOM: true, fatal: true });
+
+cachedTextDecoder.decode();
+
+let cachedUint8ArrayMemory0 = null;
+
+function getUint8ArrayMemory0() {
+    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
+        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
+    }
+    return cachedUint8ArrayMemory0;
+}
+
+function getStringFromWasm0(ptr, len) {
+    ptr = ptr >>> 0;
+    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
+}
+
+export function exported() {
+    wasm.exported();
+}
+
+export function __wbg_static_accessor_NAMESPACE_OPTIONAL_c9a4344c544120f4() {
+    const result = typeof test.NAMESPACE_OPTIONAL === 'undefined' ? null : test.NAMESPACE_OPTIONAL;
```

Comment:
This isn't correct. If the `test` namespace is not declared, this will still throw a `ReferenceError` at runtime. The code gen needs to be `typeof test === 'undefined' ? null : test.NAMESPACE_OPTIONAL`.

This also raises the question of how nested namespaces are handled. E.g. let's say I want to import the constant `A.B.C.D` (so the namespace is `A` -> `B` -> `C`). If `A` is undeclared, then the Rust value should be `None`, that's fine. But what if `B` is `undefined`? Then accessing `A.B.C.D` would result in a `TypeError` like this:

```
TypeError: Cannot read properties of undefined (reading 'C')
```

The easiest way around this would be to use [optional chaining `?.`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) (CAREFUL! `?.` does not mean the same thing as it does in Rust). Like this:

```js
const result = typeof A === 'undefined' ? null : A?.B?.C?.D;
```

However... optional chaining is relatively new (at least newer than WASM), so we can't use it. E.g. for chrome, optional chaining is supported in chrome >=80, but WBG currently supports chrome >=74 (I'm going by the [sign extension operator](https://webassembly.org/features/) here).

So we probably need to implement the optional chaining ourselves here. While not perfect, we could implement it as follows:

```js
const result = typeof A === 'undefined' || A == null || A.B == null || A.B.C == null ? null : A.B.C.D;
```

Note: In JS, `expr == null` is the same as `expr === undefined || expr === null`. It's a quick way to check for both `null` and `undefined`.

I'm saying "not perfect", because we need to access the namespaces `B` and `C` multiple times. Since JS has getters (and proxies), the value returned by `A.B` might change between accesses. So for 100% correctness, we would have to do the same thing as transpilers and store the value of `A.B` is a temporary variable. Transpilers produce code like this:

```js
let t; // temporary variable
const result = typeof A === 'undefined' || (t=A) == null || (t=t.B) == null || (t=t.C) == null ? null : t.D;
```
This is 100% equivalent to 

```js
const result = typeof A === 'undefined' ? null : A?.B?.C?.D;
```

but needs more logic on the Rust side to generate.

Reply:
I see. How about just using `try { .. } catch() { .. }`?

I should add a runtime test as well then.

Reply:
Depends. If `A.B` is a getter and throws its own error, should we ignore the error?

Reply:
> Depends. If `A.B` is a getter and throws its own error, should we ignore the error?

Probably not? I'm considering just not supporting namespaces in this case. The main use case, importing globals, would remain covered ...

Let me see how bad the implementation would be.

Reply:
Its not terrible actually.
On that note, do we want to even support getters here? `static` is thread-local cached, so if this is a getter it doesn't make sense in the first place.

So after actually implementing it ... I'm currently tending towards just not. The documentation states "static objects", not getters.

Reply:
Pushed the fix and added some runtime tests.
Let me know what you think on the getter problem.

Reply:
IMO, this is not a question of whether we support getters. People are going to assume that we do, because they won't even think about getters when they need the constant `A.B`. 

So I think it's fine to say that we don't support getters that throw in that it will cause an error, but I don't think it's fine to hide the error. Failing silently here is a problem IMO, because it hides a real bug in user code.

Reply:
I do consider our checks on bindings best effort only. From the perspective of Rust this is actually unsafe, which is why in [edition 2024 it will be written `unsafe extern "C" { ... }`](https://doc.rust-lang.org/edition-guide/rust-2024/unsafe-extern.html).

Quoting:
> Adding the `unsafe` keyword helps to emphasize that it is the responsibility of the author of the `extern` block to ensure that the signatures are correct. If the signatures are not correct, then it may result in undefined behavior.

But if we can detect that these are getters, then we should do that. I will add some documentation to that effect as well.

Reply:
> I do consider our checks on bindings best effort only.

Then we should do our best :)

`unsafe extern` isn't really relevant here, because it's not like users can use statics to express the correct signature for throwing getters. They have to use `getter` functions. (Ignore that they are broken right now.)

My point wasn't that we should support throwing getters in statics. I don't think we should. My point was that the way in which we do not support them should be better. If a getter throws for a static, it's a bug in user code either way. The question is just whether we should loudly panic and tell the user about it, or silently pretend nothing is wrong.

Note that I am assuming that JS errors thrown in the JS glue code for statics currently causes a panic. Is that correct?

Reply:
To clarify: we don't support getters at all in `static`. We cache the value thread-locally, so a using `static` with a getter is a footgun already, it was already like that.

> Note that I am assuming that JS errors thrown in the JS glue code for statics currently causes a panic. Is that correct?

No, the throw isn't caught. It will cause re-entrancy and all sort of issues, but that's a whole different rabbit-hole.

---

> > I do consider our checks on bindings best effort only.
> 
> Then we should do our best :)

What are you suggesting? *Just* detecting throwing getters is an improvement, but only a very minor one. It tells users that they are using a getter and they shouldn't. In *every other case* it will not tell users anything. I'm not happy about the code/benefit trade-off.

So I would prefer if we can detect getters to begin with. Is that possible?

Reply:
Thinking about this a bit more, I consider ignoring throwing getters an improvement actually. Because causing re-entrency is pretty bad imo.

So to actually address this, we need to do the following: catch the error, return it to Rust and panic there.

Still: ideally, we detect if its a getter in the first place.

Reply:
> So I would prefer if we can detect getters to begin with. Is that possible?

Yes, it's possible with [`Object.getOwnPropertyDescriptor`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor). 

However, it's also highly undesirable. Sometimes people use static properties, because the constant is computed and/or to cache the computation. In general, you really shouldn't differentiate between plain old property access and getters, because no one in JS land expects you to do that.

Just because something is implemented as a getter, doesn't mean that its value changes.

> What are you suggesting?

IMO, the ideal behavior would be to cause a panic in Rust. So it's unfortunate that there isn't any kind of machinery for it in place already...

My main problem with ignoring the exception is that I already know that people will start relying on it. So if we allow this behavior now, there will never be a chance to change it. So I would even prefer whatever issues not catching the exception is causing right now. At least then we get to talk about fixing it.

Also, catching and ignoring errors would be inconsistent with old plain (non-`Option<T>`) statics. So that's fun.

Reply:
> > So I would prefer if we can detect getters to begin with. Is that possible?
> 
> Yes, it's possible with [`Object.getOwnPropertyDescriptor`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor).
> 
> However, it's also highly undesirable. Sometimes people use static properties, because the constant is computed and/or to cache the computation. In general, you really shouldn't differentiate between plain old property access and getters, because no one in JS land expects you to do that.

Well, that certainly puts a wrench into my thinking ...

> > What are you suggesting?
> 
> IMO, the ideal behavior would be to cause a panic in Rust. So it's unfortunate that there isn't any kind of machinery for it in place already...

There is: `catch`. The integration is just more work I would have preferred to leave for later.

There also more design concerns with this: currently this problem concerns all bindings, not just statics. E.g. if you import a function that can throw and you don't mark it with `catch`, it will also throw in JS and may cause unexpected reentrancy into Rust.

> My main problem with ignoring the exception is that I already know that people will start relying on it. So if we allow this behavior now, there will never be a chance to change it. So I would even prefer whatever issues not catching the exception is causing right now. At least then we get to talk about fixing it.
>
> Also, catching and ignoring errors would be inconsistent with old plain (non-`Option<T>`) statics. So that's fun.

Puh, I was less concerned about it before, but if we should support getters then its definitely a concern.

---

I will go ahead then and do throw detection and adjust the documentation.

Reply:
Done.